src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -141,6 +141,36 @@ internal bool IsExceptionType(TypeSymbol type, ref HashSet<DiagnosticInfo> useSi
             return IsEqualOrDerivedFromWellKnownClass(type, WellKnownType.System_Exception, ref useSiteDiagnostics);
         }
 
+        internal bool IsWellKnownTupleType(TypeSymbol typeToCheck, int requiredArity)
+        {
+            // NOTE: error type symbol is NamedTypeSymbol, 
+            //       but not SymbolKind.NamedType, so check the kind before casting.
+            if (typeToCheck.Kind != SymbolKind.NamedType)
+            {
+                return false;
+            }
+
+            NamedTypeSymbol currentType = (NamedTypeSymbol)typeToCheck;
+            const int maxArity = TupleTypeSymbol.RestPosition;
+
+            while (true)
+            {
+                if (currentType.OriginalDefinition != this.GetWellKnownType(TupleTypeSymbol.GetTupleType(Math.Min(requiredArity, maxArity))))
```

Comment:
It feels like the code could be restructured slightly in order to avoid looking for the well known type of the maximum arity many times.